src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+
+# constant modules need to be separate from complex modules since they shouldn't mix
+module  -> ((comment|annotation):* _ "module" __ fieldName __ "{" __ (constantDcl):+ __ "}" semi) {% processConstantModule %}
+  | ((comment|annotation):* _ "module" __ fieldName __ "{" __ (structWithAnnotation|module):+ __ "}" semi) {% processComplexModule %}
+
+
+structWithAnnotation -> (comment|annotation):* struct {% 
+ // ignore annotations on structs because we can only read default value (which doesn't apply)
+ d => {return d[1];}
+%}
+
+struct -> _ "struct" __ fieldName __ "{" __ (declaration):+ __ "}" semi {% d => {
+	const name = d[3].name;
+	const definitions = d[7].flat(2).filter(def => def !== null);
+	return {
+		name,
+		definitions,
+	};
+} %}
+
+constantDcl -> (comment|annotation):* constType semi {% d => d[1] %}
+declaration -> fieldWithAnnotation semi {% d => d[0] %}
+
+fieldWithAnnotation -> annotationOrCommentLines fieldDcl {% d=> {
+	let possibleAnnotations = [];
+	if(d[0]) {
+		possibleAnnotations = d[0];
+	}
+  const fields = d[1];
+	const finalDefs = fields.map((def) => extend([...possibleAnnotations, def]));
+	return finalDefs;
+} %}
+
+fieldDcl -> (
+     _ allTypes __  multiFieldNames arrayLength _
+   | _ allTypes __ multiFieldNames _
+   | _ sequenceType __ multiFieldNames _
+ ) {% (d) => {
+	const names = d[0].splice(3, 1)[0];
+	// create a definition for each name
+	const defs = names.map((nameObj) => extend([...d[0], nameObj]));
+	return defs
+} %}
+
+multiFieldNames -> fieldName (_ "," __ fieldName):* {%
+ d => {
+	 const fieldNames = d.flat(2).filter( d => d !== null && d.name);
+	 return fieldNames;
+ }
+%}
+   
+   
+   
+annotationOrCommentLines -> (annotation|comment):* {%
+  d => {
+	  return d[0][0] ? d[0][0].filter(d => d !== null) : null;
+  }
+%}
+
+annotation -> (
+    defaultAnnotation
+  | rangeAnnotation
+  | commentAnnotation
+  | keyAnnotation
+  | transferModeAnnotation
+  ) {% d => id(id(d)) %}
+
+defaultAnnotation -> _ at "default" _ "(" _ "value" assignment _ ")" {% d => { 
+	return {defaultValue: d[7].value };
+} %}
+
+# unsupported annotations are ignored
+rangeAnnotation -> _ at "range" _ "(" _ "min" assignment _ "," _ "max" assignment _ ")" {% noop %}
+
+commentAnnotation -> _ at "verbatim" _ "(" _ "language" assignment _ "," _ "text" _ %EQ _ STR ")" {% noop %}
+
+keyAnnotation -> _ at "key" {% noop %}
+
+transferModeAnnotation -> _ at "transfer_mode" _ "(" _ %NAME _ ")" {% noop %}
+
+at -> "@" {% noop %}
+
+comment -> _ %COMMENT {% noop %}
+
+constType -> (
+     _ constKeyword __ numericType __ fieldName floatAssignment _ simple
+   | _ constKeyword __ numericType __ fieldName intAssignment _ simple
+   | _ constKeyword __ stringType __ fieldName stringAssignment _ simple
+   | _ constKeyword __ booleanType __ fieldName booleanAssignment _ simple
+) {% d => {
+	const def = extend(d[0]);
+	const name = def.name;
+	const value = def.value;
+	constantToValueMap[name] = value;
```

Comment:
is it normal to have side effects in these postprocessing functions? does that ever cause problems if the nearley engine needs to do backtracking? I think it might be more normal to just parse the AST leaving any variable/constant references intact and not worry about filling in the values until a later step.

Reply:
It technically makes it not a _context-free grammar_ but I don't see this causing any issues as is. but we can move it to a postprocessing step if you prefer

Reply:
Here's a note I found about side effects from the maintainer: https://github.com/kach/nearley/issues/383#issuecomment-400161480

Reply:
it looks like it mostly doesn't want you to modify nodes. I think this type of caching should be fine according to this comment by another maintainer further down: https://github.com/kach/nearley/issues/383#issuecomment-400638644

Reply:
I think it's fine given that we have a lexer and generally won't be partially reading in this information that will be stored, but if you feel strongly I can hoist it out and have it as a preprocessing step, thought i think that'll be a trickier implementation.